### PR TITLE
[1.17.x]: Added SendPlayerPermissionLevelEvent

### PR DESCRIPTION
**Pull request by mysticdrew**
This PR adds and event on the server for when a players permission level is changed via command or another mod. 

This is a useful event for me because I have UI that is only available for OPs and I don't want to poll the server constantly to see if the current user is an OP. With this event, I can be notified is a player is OPed via command or if another mod sets it through code, then tell the ui that the user has permissions.

### CODE REVIEW

**Comment by diesieben07** [comment]
The client is already notified of its permission level through an entity event (sent from the same method you're patching), so `LocalPlayer#permissionLevel` is always correct and thus `hasPermissions` will work fine for the client player - does this not solve your problem of not having to poll the server?

**Comment by mysticdrew** [comment]
It does for one instance, yes. 
But, we also enable/disable client features if a user is opped or not. We send a packet with the features to the client on login and dimension change as they can change per dimension. They can also be enabled for ops and disabled for non ops. This event would be beneficial so we can send the features to the client when the permission level changes. I also understand that in this case the features are enabled/disabled on the client as it is primarily a client mod and it can be modded/hacked, but it is enough for most server admins. Just like locking your car door wont stop all thieves, just most. 

I could use the command event and check for op/deop but it would not catch the case if a mod sets a users permission level through code.

**Comment by diesieben07** [comment]
Well then your initial description is wrong, since that says you don't want to poll the server from the client for the OP status, which is not necessary even without this event.

If it is added to do things on the server, the javadoc and potentially the name must be changed, because in your current implementation it fires also when the permission level has not changed at all. It fires also for example on dimension change or teleport or resource reload.

**Comment by mysticdrew** [comment]
I do have several uses for this yes, I only gave one use case I had, which I don't need this anymore for that use case. Thanks for that by the way. 
You are correct, I will update the names and javadoc.

**Comment by mysticdrew** [comment]
@diesieben07 I made the changes.

File: src/main/java/net/minecraftforge/event/entity/player/SendPlayerPermissionLevelEvent.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+package net.minecraftforge.event.entity.player;
+
+import net.minecraft.server.level.ServerPlayer;
+
+/**
+ * This event is fired on the server when a player's permission level is sent to the client.
+ *
+ * This event will fire on resource reload, when the player's dimension changes, when the player teleports, on respawn,
+ * when the new player is placed in world, and when the player is opped or deopped.
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/event/entity/player/SendPlayerPermissionLevelEvent.java]: I really wouldn't specify this that clearly here. 
> This event will fire when the player's permission level changes, but will also fire at other times when the game syncs the permission level to the client.

**Review by diesieben07 — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/event/entity/player/SendPlayerPermissionLevelEvent.java (hunk 1 above)
Comment (mysticdrew) [reply src/main/java/net/minecraftforge/event/entity/player/SendPlayerPermissionLevelEvent.java]: 👍 
Sounds good. Addressed the comment, thanks! 
Pushed

**Review by diesieben07 — APPROVED** [review]

**Comment by LexManos** [comment]
Wouldn't this be better as a 'PermissionsChangedEvent' with a old, new and player params?
So it has nothing to do with networking and is more of a 'things changed' notification.

**Comment by mysticdrew** [comment]
> Wouldn't this be better as a 'PermissionsChangedEvent' with a old, new and player params?
> So it has nothing to do with networking and is more of a 'things changed' notification.

Sure, I will update shortly.

**Comment by mysticdrew** [comment]
I updated this now only fire on op and deop and made the event cancelable.

**Review by diesieben07 — CHANGES_REQUESTED** [review]
Please use early returns in those patches to reduce the patch size.

**Comment by mysticdrew** [comment]
@diesieben07 ahh yea ok good point.

File: patches/minecraft/net/minecraft/server/players/PlayerList.java.patch [hunk 2]
```diff
@@ -81,6 +81,24 @@
...
++         return;
+       this.f_11200_.m_11381_(new ServerOpListEntry(p_11254_, this.f_11195_.m_7022_(), this.f_11200_.m_11351_(p_11254_)));
+       ServerPlayer serverplayer = this.m_11259_(p_11254_.getId());
+       if (serverplayer != null) {
+@@ -544,6 +_,8 @@
+    }
+ 
+    public void m_5750_(GameProfile p_11281_) {
++      if (!net.minecraftforge.event.ForgeEventFactory.onPermissionChanged(p_11281_, 0, this))
++         return;
```
Comment (diesieben07) [inline patches/minecraft/net/minecraft/server/players/PlayerList.java.patch]: Put the return on the same line please.

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 3]
```diff
@@ -249,7 +252,7 @@ public static Component getPlayerDisplayName(Player player, Component username)
         MinecraftForge.EVENT_BUS.post(event);
         return event.getDisplayname();
     }
-    
+
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: Unnecessary whitespace change.

**Review by diesieben07 — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/server/players/PlayerList.java.patch (hunk 2 above)
Comment (mysticdrew) [reply patches/minecraft/net/minecraft/server/players/PlayerList.java.patch]: 👍

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java (hunk 3 above)
Comment (mysticdrew) [reply src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: 👍

**Comment by mysticdrew** [comment]
Updated based on your comments @diesieben07

**Review by diesieben07 — APPROVED** [review]
